Add validation to only allow Cassandra major version 3

wallrj opened the pull request:
* Only allow versions >= 3.0.0 && < 4.0.0
* Disallow version changes in API updates (pending #277)
* Document the supported versions.

Fixes: #320
**Release note**:
```release-note
NONE
```

wallrj commented:
/retest

munnerz commented on pkg/apis/navigator/validation/cassandra.go:
```diff
@@ -1,15 +1,59 @@
+		case versionSpecGe:
+			if v.LessThan(opVersion) {
+				return false
+			}
```
Can we invert the implementation here?

i.e. `versionSpecGe` should run `v.Greater` and `v.Equal` and return true

wallrj replied:
I deleted this code and instead use github.com/hashicorp/go-version constraints.

munnerz commented on pkg/apis/navigator/validation/cassandra.go:
```diff
@@ -1,15 +1,59 @@
+			if opVersion.Equal(v) {
+				return false
+			}
+			if opVersion.LessThan(v) {
```
We switch to testing `opVersion` instead of `v` here - can we be consistent for readability?

wallrj replied:
I deleted this code and instead use github.com/hashicorp/go-version constraints.

munnerz commented on pkg/apis/navigator/validation/cassandra.go:
```diff
@@ -1,15 +1,59 @@
+	for op, opVersion := range vs {
+		out = append(out, fmt.Sprintf("%s %s", op, opVersion))
+	}
+	return strings.Join(out, ", ")
```
Maps are not consistently ordered, so this version string will be unstable. Can we make it stable?

wallrj replied:
I deleted this code and instead use github.com/hashicorp/go-version constraints.

munnerz commented on pkg/apis/navigator/validation/cassandra.go:
```diff
@@ -1,15 +1,59 @@
+	versionSpecLt: version.New("4.0.0"),
+}
+
+func (vs versionSpec) Check(v *version.Version) bool {
```
Maybe a quick unit test for this function?

wallrj replied:
Switched to github.com/hashicorp/go-version constraints which already has unit tests.

wallrj commented:
Hey @munnerz.

Please take another look.

* I replaced my home grown constraints checking code with github.com/hashicorp/go-version constraints.
* Which seems to be more popular and better tested than the alternative https://github.com/Masterminds/semver
* Also used Hashicorp version instead of coreos semver in the Cassandra Version type.
* And added generated DeepCopy functions to that, as discussed.

I can follow this up with PRs to 
 * use the same custom Version type for Elastic search.
 * have the controller perform an additional check for supported versions to handle the case where the API server may have been upgraded but the controller has not.

wallrj commented:
/retest

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -37,10 +37,28 @@ func New(s string) *Version {
+	if err != nil {
+		return err
+	}
+	v.versionString = s
```
Do we need to keep the versionString in memory? Surely `semver.String` will be more up to date?

wallrj replied:
We need it because we want to remember the format of the originally supplied version string.
So if the Cassandra node reports version `3.11` via nodetool and a JMX query, we want to report that exact version in the `Pilot.Status.Cassandra.Version` rather than the semver `3.11.0`.

We also don't want a user to submit `CassandraCluster.Spec.Version: 3.11` and then have that change to `3.11.0` when they do `kubectl get CassandraCluster -o yaml`.

However, when we infer the Docker image tag, we use the semver, because we don't want docker to simply pull the latest `3.11` image from Dockerhub.

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -2,11 +2,9 @@ package version
-	"github.com/coreos/go-semver/semver"
+	semver "github.com/hashicorp/go-version"
 )
 
```
Can we move this file out of `pkg/cassandra` and into something like `pkg/api/util`? This is an API type, but it isn't by default part of an API group (hence it isn't in apis).

It'd be good to denote it as such in the file structure, to make it clear this isn't specific to just cassandra (and also denotes that this type may be exposed in an api)

wallrj replied:
Done.

munnerz commented on pkg/cassandra/version/version_test.go:
```diff
@@ -91,3 +93,18 @@ func TestUnmarshalJSON(t *testing.T) {
+	t.Run(
+		"zero value",
+		func(t *testing.T) {
+			t.Log(version.Version{}.DeepCopy())
```
This doesn't test anything, or ever fail afaict

wallrj replied:
It tests that a Zero value can be DeepCopied.
It originally caused a panic before I added a check for nil pointer in the implementation.
I'll leave it for now.

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -49,62 +67,20 @@ func (v *Version) UnmarshalJSON(data []byte) error {
+	if v.semver == nil {
+		return Version{}
+	}
+	return *New(v.String())
```
:+1:

munnerz commented on Gopkg.toml:
```diff
@@ -52,3 +52,7 @@ required = [
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/hashicorp/go-version"
```
Don't think this needs to be added as a constraint given we are constraining it to master anyway. A simple `dep ensure` without adding this block should yield the same result, sans this change

wallrj replied:
Done.

munnerz commented:
/lgtm
/approve
